builtin/logical/transit/path_hmac.go

```diff
@@ -116,31 +128,88 @@ func (b *backend) pathHMACWrite(ctx context.Context, req *logical.Request, d *fr
 		return nil, fmt.Errorf("HMAC key value could not be computed")
 	}
 
-	var hf hash.Hash
+	var hash_alg func() hash.Hash
 	switch algorithm {
 	case "sha2-224":
-		hf = hmac.New(sha256.New224, key)
+		hash_alg = sha256.New224
 	case "sha2-256":
-		hf = hmac.New(sha256.New, key)
+		hash_alg = sha256.New
 	case "sha2-384":
-		hf = hmac.New(sha512.New384, key)
+		hash_alg = sha512.New384
 	case "sha2-512":
-		hf = hmac.New(sha512.New, key)
+		hash_alg = sha512.New
 	default:
 		p.Unlock()
 		return logical.ErrorResponse(fmt.Sprintf("unsupported algorithm %s", algorithm)), nil
 	}
-	hf.Write(input)
-	retBytes := hf.Sum(nil)
 
-	retStr := base64.StdEncoding.EncodeToString(retBytes)
-	retStr = fmt.Sprintf("vault:v%s:%s", strconv.Itoa(ver), retStr)
+	batchInputRaw := d.Raw["batch_input"]
+	var batchInputItems []BatchRequestHMACItem
+	if batchInputRaw != nil {
+		err = mapstructure.Decode(batchInputRaw, &batchInputItems)
+		if err != nil {
+			p.Unlock()
+			return nil, errwrap.Wrapf("failed to parse batch input: {{err}}", err)
+		}
+
+		if len(batchInputItems) == 0 {
+			p.Unlock()
+			return logical.ErrorResponse("missing batch input to process"), logical.ErrInvalidRequest
+		}
+	} else {
+		valueRaw, ok := d.GetOk("input")
+		if !ok {
+			p.Unlock()
+			return logical.ErrorResponse("missing input for HMAC"), logical.ErrInvalidRequest
+		}
+
+		batchInputItems = make([]BatchRequestHMACItem, 1)
+		batchInputItems[0] = BatchRequestHMACItem{
+			"input": valueRaw.(string),
+		}
+	}
+
+	batchResponseItems := make([]BatchResponseHMACItem, len(batchInputItems))
+
+	for i, item := range batchInputItems {
+		if batchResponseItems[i].Error != "" {
+			continue
+		}
+
+		rawInput, ok := item["input"]
+		if !ok {
+			batchResponseItems[i].Error = "missing input for HMAC"
+			continue
+		}
+
+		input, err := base64.StdEncoding.DecodeString(rawInput)
+		if err != nil {
+			batchResponseItems[i].Error = fmt.Sprintf("unable to decode input as base64: %s", err)
```

Comment:
I think there is a missing `continue` here

Reply:
You are correct.  

I have added another input to the test with invalid base64 data to exercise this. 

In practice, it turns out that missing the `continue` had no impact on the output and an error is generated either with or without the `continue`.  However, adding the `continue` avoids needlessly generating an HMAC of the empty input string - so let's do that.